### ISSUE TITLE
thought experiment: breaking with the past

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#6602]
# breaking with the past for borg 2.0

what could be done if we decide to make a breaking release that:
- does not try to be compatible with old repos / old cli syntax
- only uses new repos / keys it created itself
- only gets old archives via import-tar

## crypto

- remove all old crypto
- remove pbkdf2 + pbkdfs/sha256 keys + docs - we have argon2 now which is better anyway
- remove key upgrade / downgrade code (pity, still fresh) - not needed if we break with the past
- remove low_level.AES class for pbkdf2 key encryption (argon2 keys use something more modern)
- aes-ctr mode + all docs talking about nonce reuse and related - first close #6601
- nonce management code for aes-ctr, not needed any more with session keys
- remove nonces module
- remove support for super-legacy passphrase-key type (not supported since long)
- if we adopt blake3, we could even remove blake2 based stuff

## repo

- remove PUT support, only have PUT2
- remove detection of attic repos
- remove free nonce / nonce reservation api
- remove hints v1
- optimize with read_data=False, it now checks the crc32
- if we reduce MAX_OBJECT_SIZE from ~20MiB to 16.000.000, 24bit are enough for the entry length
   - better alignment for segment entries together with the 1 type byte.
   - in-memory indexes have 8 free bits without using more memory
   - max archive size goes down by 20%!
   - do not allow 16MiB objects, we need some room for potential header size increase in future

## indexes / cache

- repo index with entry size / csize (no random accesses needed for quota / hints updates when deleting)
- set support in index? just some bits one can set/clear/query + set operations.
  - if we get the csize down to 24bits, we have 8 free bits!
  - we could use a few of these bits to remove the in-band signaling for the hashtable internal purposes (used vs free bucket, tombstone, ...).
- see `legacy_cleanup` function

## msgpack

- switch to more modern msgpack types (for str vs bytes)
- be super careful with PAX-BORG, uses borg 2.0 code for import-tar and borg 1.x for export-tar
  (idea: implement the 1.x export-tar with the modern msgpack params)!
- check the bigint stuff #2323, timestamp 96 msgpack type, timestamp support of msgpack-python in general

## archive / item

- do not support mtime-only items in archives
- always have borg12_meta in new archives, remove support cache for old archives
- resolve dual use of item.source (hardlink and symlink)
- get rid of csize in item chunk list
- remove stoneage attic bug support in valid_item (erroneous b'acl' key)
- get rid of hardlink_master and complex code dealing with it, tends to cause issues. #855

## checksums

- remove libdeflate again and use zlib.crc32 from stdlib, PUT2 format only uses crc32 for header data, no big stuff
- guess we do not want to replace header crc32 by xxh64? different properties crcs vs. hashes.

## compression

- remove lzma and zlib, lz4 + zstd is better anyway
- in case we keep zlib, remove the dirty hack for the type bytes

## upgrade

- remove upgrades from old borg

## archiver

- remove unneeded stuff from benchmark cpu
- recreate vs tar-export/tar-import? do we need both?

## remote

- drop support of all non-current rpc protocol formats

## cli

- separate archive name from repo url (our regexes are way too complex)
- drop scp url syntax (problematic, can not use port spec, regex tends to cause bad matches)?

## general comments

putting all the breaking stuff into 1 release is good for users (1 time effort),
but will take quite some time to test and release.

**Comment by elho** [comment]
I do not mind breaking for the better at all, but some of the outlined details do not qualify for that IMHO.

When it comes to crypto, breakage should not occur to replace one algorithm with a limited life span with another one with a limited life span and thus planning with breakage every few years. Instead breakage should be done to end up with a repo format that does support multiple algorithms and easy and feasable changing of keys as well as used algorithms. That could e.g. be by at least temporarily allowing multiple algorithms to be "active" in a repo at the same time.

When it comes to repo format, a breakage should not be the excuse to just dump a bit of code to still support reading PUTs besides PUT2s, but question the format as a whole and try to address issues such as the current limitations of append-only as well as secure multi-client usage, infeasible (with huge repos) compaction.
Ideas here would be:
- split into server managed (optionally *hard* add-only) content chunk "pool" and a per-client (separate crypto) meta-data stores (that could or could not be chunk based).
- consider use of public-key crypto especially in such a split storage model, where clients do not have the secret key that would be needed to read back chunks from the content pool, but only the one for the meta-data store, with the meta-data being encrypted to both that one and one of the borg server for the meta-data, so that it can both use the meta-data for e.g. compaction of the content chunk pool and auditing of client behavior (limit access to content chunks to those the client does have "written" before). The devil is in the details, but again, just ideas.
- evaluate radically repo formats, e.g. hierarchy of nested directories plus files named by chunk-id (prefix), leading in the extreme to a segment-less format where chunk existence is a single stat, chunk deletion a single rm of a single file and reading a chunk would be simply entirely reading a single file, or in less extreme cases (should practical testing reveal that filesystems would be a limiting factor) to segments that are not filled one after another, but collecting chunks with the same prefix. The extreme case would on the plus side lend itself perfectly to object storage backends, which a people keep asking for. (Both these could turn out to be terrible ideas, but they and other wild ideas should be considered and looked into to make sure a completely breaking repo format is one to be kept for a long time and worth the breakage)

When it comes to compression, what really should go is the auto mode - or be reimplemented with useful parameters, whcih IMO are hard to come up with in the light of ZSTD performance.
While I agree that zlib is of the past and LZ4 could be replaced by low ZSTD modes, I strongly oppose to dropping LZMA which [in my experience](http://www.elho.net/dev/borg/#compression) is still able to produce smaller results than ZSTD. In a world where for remote backups bandwith often is by far the main bottleneck, CPUs still getting faster and the hope for an asynchronously operating borg (that then could easily be fitted with multiple compressor threads) not being entirely dead (or is it?), it in my book makes no sense to dismiss strong compression. Also there's data-sets that see little change, so in a long-term perspective, the speed of one time initial compression hardly matters.

By "scp URL" I assume you mean the scp/rsync/etc. syntax that is **not** a URL and instead enforce URL only (though what about local paths, `file://` only, too?)
On the one hand I think it does not matter much, any sane setup does have wrapper scripts around it to make you only ever see and use the repo URL once in the life of the repo.
On the other hand, given the use in scp/rysnc etc. maknig that non-URL syntax so much more common to users and thinknig how obnoxious `file://` URLs are when combined with UNIX paths starting with another slash, plus that while the code handling things leaves a lot room for improvement, a lot of that has nothing to do with the non-URL syntax as such.

**Comment by ThomasWaldmann** [comment]
Crypto:

AES-CTR does not have a limited timespan. Why we are doing this is to get rid of the fundamental counter management issues:
- you can't lose the local counter memory and not trust (and continue to use) the remote repo.
- also you can't use multiple clients for one repo and not trust the repo.

There's also a slight ugliness of only storing a part of the IV within the old format, but that is just a minor detail.

The new AEAD algorithms with session keys solve that.

We could have all 3 crypto algorithms in parallel in the borg code (but currently not in same repo), but there are other things on the above list that are best solved with tar-export/import and a new repo and IF ones does that anyway, one can as well go for the better crypto in one go (instead of having to do the export/import again some time later).

I don't think it would be a good idea to use different encryption algorithms in the same repo and especially not with the same key - so if we would go for the complexity of supporting repos with that, we would need multiple (master) keys for one repo, making it more complex for borg and also for the users.

You also can't just "change the keys / algos" in the same repo. Due to dedup, a lot of data would be still encrypted by old key and old algorithm. To get really rid of it you'ld need some global migration, touching a lot of data and needing some management for the case of interruptions of that process. That's about as much I/O and time needed as the export/import, just with much more complexity.

**Comment by ThomasWaldmann** [comment]
Repository:

It's not just about the "reading PUTs" - it is at quite some places, including borg check (which is already quite complex).

I can imagine doing some more and even radical changes to the repo format if we re-start with new repos and require export/import anyway. I am not too happy with the complexities of segment file handling either.

In the end this will depend on some developers architecting and implementing it though and we should try to not make the scope too big though or it'll never get releasable.

Repos: interesting ideas. Needs more analysis I guess, esp. since we likely want to keep the transactional behaviour and maybe also the LOG like behaviour.

Segmentless repos: if everybody had a great repo filesystem and enough storage, I guess that could be done (but it would mean that if the source has a million files, the repo could have XX million chunks). Super simple for borg, but a huge load on the repo fs (did that within my zborg experiment back then). Could also be quite slower due to more random accesses and more file opening and use a lot more space due to fs allocation overheads if one has a significant amount of small files.

Cloud storage: I don't want to maintain such code myself, that's just a rabbit hole I don't want to get into. So, for me it is "local directory" as the repo (plus some method of remoting that, not necessarily the hard to debug current remote.py code).

**Comment by ThomasWaldmann** [comment]
Compression:

I didn't want to replace lz4 with zstd. Rather zlib and lzma replaced by zstd and keeping lz4 as fastest method.

I see in your measurement that lzma was 3% better than zstd (but also even slower - in general, these last few percents of compression are paid with a high price in cpu cycles). There is also that other ticket about lzma ...

Yes, I agree that we get more and faster CPU cores and we will have multithreading in borg at some time, but storage volumes and transmission speeds are increasing also.

**Comment by ThomasWaldmann** [comment]
TARGET param/URL/syntax:

- i would not like to require file:// because the official spec is abysmal and also it should be the default (as a normal path).
- yeah, I meant scp syntax user@host:/path - borg has a quite complex regex for that **which is also used for local paths**. people sometimes forget the colon and the regex gets fixes every few years because it did not work correctly and it is always a pain.
- ssh://user@host:port/... is not that much effort and cleanly separated from local repos. if one needs it frequently on the command line, one would set BORG_REPO anyway.

**Comment by ThomasWaldmann** [comment]
@elho thanks for the detailled feedback btw!

This ticket is primarily meant for the to-break-or-not-to-break decision. Once we decide to do a breaking release, requiring new repos, key, export/import, we can do a lot of changes and need to discuss the details in more specific tickets.

We should somehow try to limit the scope though, so it won't take forever.

**Comment by RonnyPfannschmidt** [comment]
@ThomasWaldmann if instead of segments something like git pack's could be used, then with the new encryption session stuff it may even turn feasible to push packs instead of archives between repos without necessarily requiring de/encryption

**Comment by RonnyPfannschmidt** [comment]
Potentially this would also enable potentially dumb remotes like s3, sshfs, with the caveat of having more pain with post prune gc and repacking

**Comment by ThomasWaldmann** [comment]
@RonnyPfannschmidt encrypted chunks can be transferred between related repos using the same key material, there is a ticket about that already. I don't know the git pack format, so not sure how that is relevant for (re-)encrypting. But if we want to transfer a full "pack", there might be requirements due to that (opposed to just transferring a single chunk).

**Comment by elho** [comment]
> It's not just about the "reading PUTs" - it is at quite some places, including borg check (which is already quite complex).
What I am after is an upgrade path, not compatibility to the extent of fully operationg the previous repo format. So I would expect a borg1.3 check on a v1 repo to show an error of unsupported repo format.

I would be happy with a borg1.3 that on first use of serve on (or direct local access to) a v1 repo would start out (maybe after some confirmation) by iterating over all segments, for each creating a new replacement segment file, filling it with the same content except for using PUT2 whenever a PUT is read from the old one, doing some sort of verify pass maknig sure the new segment as arrived on disk has the same data as the old one and only then atomically mv the new over the old one. When having done the last segment file without being interrupted, switch repo version from v1 to v2.
No other command or code path would need to support v1 and PUT in that scenario.

> I didn't want to replace lz4
Ah, I parsed that wrong then. Still, something in the range of zstd,1 to zstd,3 may easily beat lz4 in both speed and size, which may have contributed to me parsing it that way. :wink:

> Why we are doing this is to get rid of the fundamental counter management issues
To clarify that as well, will/may there remain some other AES-SOMETHINGTHATISNOTCTR mode, or is it as I parsed that part, that eventually blake3 could be the only supported algorithm?

**Comment by horazont** [comment]
> recreate vs tar-export/tar-import? do we need both?

I'd say, yes.

`recreate` has the advantage of not needing an entire archive's worth of scratch space when changing an archive (or a bunch of archives) in a repository. AIUI, with tar-{export,import}, you'd have to store the entire archive as tar archive somewhere if you wanted to modify an archive within the same repo (because locking, otherwise you could `borg export-tar | filter-somehow | borg import-tar` probably).

Requiring an entire archive worth of scratch space is probably a deal-breaker in many non-professional and some professional settings.

tar-{export,import} is required though for moving archives between repos, so I'd say both are in fact needed.

**Comment by ThomasWaldmann** [comment]
@horazont agreed, strike-through markup in my list.

**Comment by RonnyPfannschmidt** [comment]
I don't see how a scratch space would be needed as one can stream the archives

But not having to tarify/rechunkify/recommpres/reencrypt on "filtered Reimport" is still nice to have

**Comment by ThomasWaldmann** [comment]
@RonnyPfannschmidt only 1 client in a repo at a time (locking) makes a temp tar file necessary if archive source and destination is the same repo (e.g. if wanting to change compression or chunking, not moving to another repo).

**Comment by RonnyPfannschmidt** [comment]
@ThomasWaldmann would it be feasible to distinguish between read, append and change locks?

**Comment by ThomasWaldmann** [comment]
Maybe? I am not too motivated to look into that now though.

Needs a lot of review and we already have a more differentiate locking system, including deadlocks.

**Comment by ThomasWaldmann** [comment]
Note: I updated the topmost post with feedback from you all (thanks!) and also with new insights. I also edited some other posts to remove duplicate / outdated information to keep this issue short.

**Comment by ThomasWaldmann** [comment]
Progress in #6663 and #6668 looks quite good.

About version: if we require people to transfer their repos using `borg transfer`, guess that must be borg 2.0 because you can't just continue with an existing repo as it is.

**Comment by horazont** [comment]
> not sure if we can already do that. a lot of platforms already dropped 32bit support, but for some this is still in the works (e.g. SBC like the raspberry pi).

I think especially SBCs will stay 32bit for a while, because the savings in having a smaller pointer width are relevant on low-memory platforms.

Aren't there clock system calls which return a 64-bit wide integer even on 32-bit ABIs?

**Comment by ThomasWaldmann** [comment]
Well, it's not just like borg needs to get the 64bit time by doing a call, it rather is the whole system of kernel / libc / python needing to work with timestamps of reasonable length. E.g. timestamps in `os.stat` output, python `time` and `datetime` stuff, etc.

**Comment by elho** [comment]
> So, if we merge these, next release from master will not be 1.3, but 2.0.

Changing the module name from `borg` to `borg2` at this point is something to be thoroughly considered.

Both, to eventually play with potential (meanwhile obsoleted already) export/import tar magic, but also to be able to test 1.2 in parallel with 1.1 in production across all my systems in a sane manner, I went on the surprisingly painful adventure to create myself a variant of the distribution's package that can be installed and used in parallel with the stock 1.1 one. In a hackish manner, one could install borg below a different path, but that is nothing any distribution would do, I went the painful way to do such a rename in there.
(IOW happy to clean that up and even break out some of the cases where absolute imports were used without need and against the common practice in most other similar places in the code).

For the original idea of export-import migration this would be a requirement, here it is not, but in practice, for people backing up to multiple repos, scenarios like migrating the local one to 2.0 while still waiting an undefined time for the borg storage provider the external one resides on to support 2.0 could be very common.

**Comment by ThomasWaldmann** [comment]
Guess it is not just about the module name, but also the cli cmd name. OTOH, I'ld dislike to put the version number into the cli cmd name.

For testing, one could also use the fat binary and rename that to borg2.

**Comment by elho** [comment]
> Guess it is not just about the module name, but also the cli cmd name. OTOH, I'ld dislike to put the version number into the cli cmd name.

It is, but the command name is something that can just be changed without requiring any modification of the command itself to keep it working, and on the other hand is something distributions have support for.
E.g. in Debian, a borg2 package would ship `borg2` etc. comnands, but (along with a packaging update to the 1.x version to be shipped in parallel) make use of the `alternatives` system of managed symlinks to have `borg` commands available to the user that point to whichever version is installed on its own, to (probably best for compatibility) `borg1` if both are installed, with the option for the user to easy switch that (along with the corresponding manpages) according to his preference.

Aware wrappers that censequently have an idea of the configured repo(s) being version 1 or 2 would know to invoke according versioned command name in all cases.

> For testing, one could also use the fat binary and rename that to borg2.

Testing as in "is this for me" or "does this work at all", yes. But not for testing as in "let me run this in parallel to 1.1 for a couple months and see whether any issues arise before ditching 1.1", ie. a point where 1.2 can be regarded to be at currently.

**Comment by horazont** [comment]
> Well, it's not just like borg needs to get the 64bit time by doing a call, it rather is the whole system of kernel / libc / python needing to work with timestamps of reasonable length. E.g. timestamps in `os.stat` output, python `time` and `datetime` stuff, etc.

The `statx` syscall already has 64-bit wide timestamps (it uses `__s64` for the seconds instead of `time_t`). [Since kernel 5.1, 64-bit wide time structs are available on a bunch of other system calls](https://lwn.net/Articles/776435/). 

So the kernel can (probably; I saw patches for utimes64, not sure if those have been applied, it hasn't been mentioned in that post above) do it.

I'm not sure what the [current status is on the glibc side of things](https://sourceware.org/glibc/wiki/Y2038ProofnessDesign#General_implementation) (the page looks a bit unclear on progress), but it may be worth pushing python on 32bit architectures to use it if glibc is ready.

All I'm saying: don't drop support for 32-bit architectures, but go for dropping support for 32-bit timestamps, which don't have to be the same thing anymore this time and age.

**Comment by fantasya-pbem** [comment]
I'd like to suggest an enhancement of return codes / error codes. We currently only have 1 for warning and 2 for error (and 128+N if killed by signal N). We should at least reserve/define a broader range of codes that could be implemented in future versions, for example:

1-63: warning
64-127: error
128+N: (unchanged)

If we don't want to implement more detailed return values yet, at least we would change rc 2 to 64. Users would have to change their wrappers from rc 2 to rc >= 64 to recognize errors and could keep the logic for warning.

If we want, we could already define some sub-ranges, e.g. "network errors", "disk related", "quota related"...

**Comment by ThomasWaldmann** [comment]
IIRC there is a separate ticket already for that, please move that there.

**Comment by fantasya-pbem** [comment]
Sorry, but I didn't find a matching issue yet. This ticket is about the "decision to break or not to break" (and the decision seems to be "break it"), so I wanted to add the error code topic here. Do you want me to create a new issue for this?

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem return codes: I added #6756 to the list in the top post. please move all relevant information there and delete posts from this ticket.

**Comment by ThomasWaldmann** [comment]
Note: i update the top post with the current progress and also released 2.0.0a3 - if no one is holding me back with negative testing results, I'll soon merge the `borg2` branch into `master`.

**Comment by ThomasWaldmann** [comment]
@FabioPedretti the point of using zlib is that python3 provides it, so there is no additional dependency for borg.

**Comment by ThomasWaldmann** [comment]
as there was no negative feedback from alpha testing, i just merged the `borg2` branch into `master`. 🚀

keeping this issue open until N+1 for the misc. remaining TODO.

**Comment by xeruf** [comment]
Is there any overview of what borg2 improves for me as a user? How usable is it?

**Comment by ThomasWaldmann** [comment]
IIRC I did not write a short overview yet, so there's what you can read in the change log and in the top post of this ticket.

The super short overview is "we fixed most issues labelled as BREAKING", they often were long-term open issues (sometimes since attic) because fixing them breaks compatibility.

2.0.0b1 should be pretty usable, just do not run it against production repos (rather use copies to experiment).

**Comment by ThomasWaldmann** [comment]
@xeruf see there.

**Comment by RubenKelevra** [comment]
How about using zstd dictionaries to get the compression ratio up? :)

**Comment by ThomasWaldmann** [comment]
@RubenKelevra do you have an idea about how exactly would that work inside borg?

**Comment by RubenKelevra** [comment]
@ThomasWaldmann sure:

- Run a bunch of test files through `file` to determine the mime type.
- Cut them in pieces with buzhash
- Save the pieces on the disk in folders named like the mime types
- Then let zstd create a dictionary for each mime type.
- Create a table with byte size numbers for each mime type


Rationale behind the last step is: zstd archives can select the used dictionary for decompression by a byte (as an minimum size identifier). Since on the block level it's probably pretty tricky to get the mime type before decompressing the file, it's probably best to let zstd choose the correct dictionary by an identifier stored in each block (takes up one byte).

This becomes important if the same data is found in different types of files. Say a tar archive contains blocks of a JSON file.

The mime type is in this case no longer helpful, but decompression is still possible.

**Comment by ThomasWaldmann** [comment]
@RubenKelevra well, I see what you mean, but that is not how "borg create" works.

**Comment by RubenKelevra** [comment]
> @RubenKelevra well, I see what you mean, but that is not how "borg create" works.

Interesting, can you elaborate or point me to the part which is different than I think, so I can take a look? 🤔 


> But maybe check the issue tracker if we have a ticket about this and if not, create a new one, so we can collect ideas there.

Will do

**Comment by neuhaus** [comment]
Hello, i am a new user of borg. I need a usable backup by the end of December.
Is the release of Borg 2.0 planned by the end of this year? If so i could start with it.

**Comment by ThomasWaldmann** [comment]
I can't promise that.

And nobody should delay having a backup due to waiting for a release, just use the current stable version.

**Comment by AlphaJack** [comment]
Any plan to implement multithreading #37 in 2.0?

**Comment by ThomasWaldmann** [comment]
No multithreading in 2.0, that will be in a later release.

**Comment by arodland** [comment]
There shouldn't be any need to drop 32-bit support to be y2038-clean. 32-bit platforms can still have a 64-bit time_t, and most of them do, and have done for 5-10 years at least.

**Comment by enkore** [comment]
Have there been any major complaints / pain points with the JSON API? The only things I've found is

(a) (largely hypothetical) encoding woes when involving file names (obviously file names don't have to be representable in unicode regardless of locale) and on weird systems (#2273) and
(b) it's annoying to parse when stdout and stderr are multiplexed, because stdout uses pretty printing (#6053, #3605)

**Comment by ThomasWaldmann** [comment]
@enkore the json encoding issues for e.g. `path` and also some other things that can not be represented as valid unicode (== without surrogate escapes) were solved some months ago, e.g.:

- have `path` member with an unicode approximation ("printable")
- have `path_b64` with an exact base64 representation of the original bytes

Especially on samba servers this is not at all hypothetical, but a very practical issue, because the servers existing since some decades already collected all sorts of historical `path` encodings.

**Comment by FabioPedretti** [comment]
1. A minor issue in doc for 2.0.0b6: the line `Archives compressed with zstd are not compatible with borg < 1.1.4.` can be removed, since borg2 no longer supports old repos anyway, correct?

```
$ borg-2.0.0b6 help compression
[...]
zstd[,L]
    Use zstd ("zstandard") compression, a modern wide-range algorithm.
    If you do not explicitly give the compression level L (ranging from 1
    to 22), it will use level 3.
    Archives compressed with zstd are not compatible with borg < 1.1.4.
[...]
```

2. And another consideration: maybe zlib compression can be removed altogether, since zstd already covers all its use cases (zstd can be faster and higher compression at the same time than zlib, even if using a bit more memory)?

**Comment by ThomasWaldmann** [comment]
@FabioPedretti : 

1. good catch, fixed by #7800.
2. yeah, but the `borg transfer` archive copying used to migrate from borg 1 to borg 2 repos does not re-compress for efficiency reasons, so borg 2 must support all borg 1 compression algorithms.

**Comment by FabioPedretti** [comment]
Thanks, about zlib, OK for keeping it for borg transfer (and decompression), maybe just disable it when requesting it for compressing new archives? Leaving it may also be OK (since it won't clean up a lot of code), it could eventually be used for some unknown use case?

**Comment by ThomasWaldmann** [comment]
I don't really see a point of disabling it in borg2. It's not the default, so users can decide to use whatever they want.

And even if we disabled it in borg2, users could still have old archives with zlib compression in borg2 repos due to `borg transfer`.

**Comment by RafaelKr** [comment]
@ThomasWaldmann Looking at your OP nearly everything is marked as done 🥳 Is there still a lot to do before a stable release or is it now in the final testing phase?

**Comment by ThomasWaldmann** [comment]
@RafaelKr can't say that yet. I am thinking about whether I should do more (and very fundamental) changes now (avoiding having to do another breaking release in future, but delaying borg2 a lot) or rather release it sooner as planned above.

**Comment by issmirnov** [comment]
@ThomasWaldmann my vote is on delaying the release and only doing one breaking change. Otherwise, your users will have to migrate v1-v2 with breaking changes, and then within a "short" time (6-12 months?) have to migrate v2-v3. Some users will be on v1, so you'd also have to build out v1-v3 upgrade paths and checks.

Borg v1 works great, we've waited this long, we can wait a little longer to just have to pay the pain of migration once. 

Everyone, feel free to thumbs up / thumbs down this comment to express your opinion.

**Comment by tmm360** [comment]
I think that all is a matter of timing. How much is "a lot"? If is 6 months, merge them. If it is a fundamental rewrite and will take 2 or more years to be stable, do two separate releases.

**Comment by knutov** [comment]
my two cents: if it's ready - it's ready.

some changes will happen eventually, there is no problem to do small updates in scripts.

New version has a lot of benefits, why wait to us it?

**Comment by ThomasWaldmann** [comment]
@tmm360 What I have in mind is a big change (not even sure how big), my and other contributors' free time is a bit hard to predict, so it makes the overall time needed somehow unpredictable.

Maybe forking off some new borg-ng branch from master and just starting that development there, while fixing bugs and missing stuff in master branch would be an option. Depending on more insights developing over time, a release could be made from either branch.

**Comment by tmm360** [comment]
@ThomasWaldmann at this point I've no doubt it should be another release, and keep time to develop it without need to hurry. It looks something of huge, and if borg2 is ready, my idea it should be released as is.

**Comment by darkk** [comment]
Speaking of pro and contra I'd also add that migration might also have a _desirable_ side-effect of backup verification.

However, I understand that the process might require twice as much of storage space under certain conditions.

**Comment by j1warren** [comment]
was 2.0 put on hold?

**Comment by RafaelKr** [comment]
> was 2.0 put on hold?

The commit history tells it is actively worked on: https://github.com/borgbackup/borg/commits/master/

**Comment by ThomasWaldmann** [comment]
@j1warren As my work on borg2 will likely take quite a bit longer, I temporarily switched focus to borg 1.x and made a "refresh" there in form of borg 1.4 (currently beta 1), see #7975 for more details. Will continue to work on borg2 soon.

**Comment by struthio** [comment]
@ThomasWaldmann How usable branch 2.0 currently is ? If I start new repo now, will it be usable with final 2.0 version or I will have to start from scratch when final 2.0 is released ?
I am going to start new repo for very big dataset and I would like to avoid migration in the future.

**Comment by ThomasWaldmann** [comment]
@struthio borg 2.0 is NOT for production and there is no compatibility or upgrade procedure expectation between betas / RCs or later a release.

If you need something stable, use 1.2.7 (or soon 1.4.x).

**Comment by struthio** [comment]
> upgrade procedure expectation between betas / RCs or later a release.
Ok that's clear. Thanks.

**Comment by mibmo** [comment]
Is there any _eta_ for when borg 2 will release? despite clear recommendations, judging by this issue it seems to be in good shape (although the issue hasn't been updated since 2023?) :)

**Comment by ThomasWaldmann** [comment]
@mibmo No ETA, see also https://github.com/borgbackup/borg/issues/6602#issuecomment-1911993485.

**Comment by struthio** [comment]
> The borg 2.0 code will still need to deal with reading borg 1.x archives for borg transfer to migrate them into borg 2 repo, thus we have to be a bit careful not to tear down some stuff we still need.

Can it be released in two steps.
Step 1. Borg2 with new archive format etc is ready, but only for new repositories.
Step 2. Compatibility with borg1 archives

**Comment by ThomasWaldmann** [comment]
@struthio well, the compatiblity code is just the old code, so it is already present. And unit tests say, that borg transfer works in my #8332 PR, so no need for 2 steps.

**Comment by ThomasWaldmann** [comment]
#8332 was merged into master.

everybody can help beta testing this huge change in 2.0.0b10, ETA is tomorrow.

**Comment by ThomasWaldmann** [comment]
Via the borgstore rclone backend, borg just got cloud storage support.

**Comment by drybx** [comment]
Out of curiosity:
Is there already a timeline on a release of version 2.0?
Will it include the latest version of Zstandard?
Thanks!

**Comment by ThomasWaldmann** [comment]
@drybx No, still in beta, still a lot to do.

borg 1.4 branch and also master branch do not bundle the source code of zstd anymore, but they use the zstd library/headers installed on the system. That's better for dist packaging and also for security or other bug fixes.

**Comment by ThomasWaldmann** [comment]
Some people were wondering about what is still todo for 2.0.

It is mainly that:

https://github.com/borgbackup/borg/issues?q=state%3Aopen%20label%3A%22breaking%22

**Comment by ThomasWaldmann** [comment]
Packs will be in 2.0.0b22 (soon).
